File: app/src/main/java/org/monksanctum/xand11/graphics/GraphicsContext.java
// Copyright 2018 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package org.monksanctum.xand11.graphics;

import android.graphics.Paint;
import android.graphics.Paint.Cap;
import android.graphics.Paint.Join;
import android.graphics.Xfermode;

import org.monksanctum.xand11.errors.ValueError;

public class GraphicsContext {

    private final int mId;

    public int drawable;
    public byte function;
    public int planeMask;
    public int foreground = 0xff000000;
    public int background = 0xffffffff;
    public int lineWidth; // Card16
    public byte lineStyle;
    public byte capStyle;
    public byte joinStyle;
    public byte fillStyle;
    public byte fillRule;
    public int tile; // PIXMAP
    public int stipple; // PIXMAP
    public int tileStippleX; // Card16
    public int tileStippleY; // Card16
    public int font; // Font
    public byte subwindowMode;
    public boolean graphicsExposures;
    public int clipX; // Card16
    public int clipY; // Card16
    public int clipMask; // PIXMAP
    public int dashOffset; // Card16
    public byte dashes;
    public byte arcMode;

    private Paint mPaint;

    public GraphicsContext(int id) {
        mId = id;
    }

    public void createPaint() throws ValueError {
        mPaint = new Paint();
        mPaint.setColor(foreground | 0xff000000);
        mPaint.setStrokeWidth(lineWidth);
        switch (function) {
            case FUNCTION_XOR:
                try {
                    mPaint.setXfermode((Xfermode) Class.forName("android.graphics.PixelXorXfermode")
                            .getConstructor(int.class).newInstance(0xffffffff));
                } catch (Exception e) {
                    e.printStackTrace();
                }
                break;
            default:
                mPaint.setXfermode(null);
                break;
        }
        switch (capStyle) {
            case CAP_STYLE_NOT_LAST:
            case CAP_STYLE_BUTT:
                mPaint.setStrokeCap(Cap.BUTT);
                break;
            case CAP_STYLE_ROUND:
                mPaint.setStrokeCap(Cap.ROUND);
                break;
            case CAP_STYLE_PROJECTING:
                mPaint.setStrokeCap(Cap.SQUARE);
                break;
        }
        switch (joinStyle) {
            case JOIN_STYLE_MITER:
                mPaint.setStrokeJoin(Join.MITER);
                break;
            case JOIN_STYLE_ROUND:
                mPaint.setStrokeJoin(Join.ROUND);
                break;
            case JOIN_STYLE_BEVEL:
                mPaint.setStrokeJoin(Join.BEVEL);
                break;
        }
    }

    public static final byte FUNCTION_CLEAR = 0;
    public static final byte FUNCTION_AND = 1;
    public static final byte FUNCTION_AND_REVERSE = 2;
    public static final byte FUNCTION_COPY = 3;
    public static final byte FUNCTION_AND_INVERTED = 4;
    public static final byte FUNCTION_NOOP = 5;
    public static final byte FUNCTION_XOR = 6;
    public static final byte FUNCTION_OR = 7;
    public static final byte FUNCTION_NOR = 8;
    public static final byte FUNCTION_EQUIV = 9;
    public static final byte FUNCTION_INVERT = 10;
    public static final byte FUNCTION_OR_REVERSE = 11;
    public static final byte FUNCTION_COPY_INVERTED = 12;
    public static final byte FUNCTION_OR_INVERTED = 13;
    public static final byte FUNCTION_NAND = 14;
    public static final byte FUNCTION_SET = 15;

    public static final byte STYLE_SOLID = 0;
    public static final byte STYLE_ON_OFF_DASH = 1;
    public static final byte STYLE_DOUBLE_DASH = 2;

    public static final byte CAP_STYLE_NOT_LAST = 0;
    public static final byte CAP_STYLE_BUTT = 1;
    public static final byte CAP_STYLE_ROUND = 2;
    public static final byte CAP_STYLE_PROJECTING = 3;

    public static final byte JOIN_STYLE_MITER = 0;
    public static final byte JOIN_STYLE_ROUND = 1;
    public static final byte JOIN_STYLE_BEVEL = 2;

    public static final byte FILL_STYLE_SOLID = 0;
    public static final byte FILL_STYLE_TILED = 1;
    public static final byte FILL_STYLE_STIPPLED = 2;
    public static final byte FILL_STYLE_OPAQUE_STIPPLED = 3;

    public static final byte FILL_RULE_EVEN_ODD = 0;
    public static final byte FILL_RULE_WINDING = 1;

    public static final byte SUBWINDOW_MODE_CLIP_BY_CHILDREN = 0;
    public static final byte SUBWINDOW_MODE_INCLUDE_INFERIORS = 1;

    public static final byte ARC_MODE_CHORD = 0;
    public static final byte ARC_MODE_PIE_SLICE = 1;

    public Paint getPaint() {
        return mPaint;
    }
}

File: app/src/main/java/org/monksanctum/xand11/graphics/GraphicsContext.kt
// Copyright 2018 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package org.monksanctum.xand11.graphics

import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Paint.Cap
import android.graphics.Paint.Join
import android.graphics.Path
import android.graphics.Xfermode

import org.monksanctum.xand11.errors.ValueError
import org.monksanctum.xand11.fonts.Font
import org.monksanctum.xand11.fonts.FontManager

class GraphicsContext(private val mId: Int) {

    var drawable: Int = 0
    var function: Byte = 0
    var planeMask: Int = 0
    var foreground = -0x1000000
    var background = -0x1
    var lineWidth: Int = 0 // Card16
    var lineStyle: Byte = 0
    var capStyle: Byte = 0
    var joinStyle: Byte = 0
    var fillStyle: Byte = 0
    var fillRule: Byte = 0
    var tile: Int = 0 // PIXMAP
    var stipple: Int = 0 // PIXMAP
    var tileStippleX: Int = 0 // Card16
    var tileStippleY: Int = 0 // Card16
    var font: Int = 0 // Font
    var subwindowMode: Byte = 0
    var graphicsExposures: Boolean = false
    var clipX: Int = 0 // Card16
    var clipY: Int = 0 // Card16
    var clipMask: Int = 0 // PIXMAP
    var dashOffset: Int = 0 // Card16
    var dashes: Byte = 0
    var arcMode: Byte = 0

    var paint: Paint? = null
        private set
    private var p: Path? = null

    fun applyToPaint(p: Paint): Paint {
        p.color = foreground or -0x1000000
        p.strokeWidth = lineWidth.toFloat()
        when (function) {
            FUNCTION_XOR -> {
            }
            else -> p.xfermode = null
        }// TODO: Support this.
        when (capStyle) {
            CAP_STYLE_NOT_LAST, CAP_STYLE_BUTT -> p.strokeCap = Cap.BUTT
            CAP_STYLE_ROUND -> p.strokeCap = Cap.ROUND
            CAP_STYLE_PROJECTING -> p.strokeCap = Cap.SQUARE
        }
        when (joinStyle) {
            JOIN_STYLE_MITER -> p.strokeJoin = Join.MITER
            JOIN_STYLE_ROUND -> p.strokeJoin = Join.ROUND
            JOIN_STYLE_BEVEL -> p.strokeJoin = Join.BEVEL
        }
        return p
    }

    @Throws(ValueError::class)
    fun createPaint(fontManager: FontManager) {
        val f = fontManager.getFont(font)
        paint = applyToPaint(if (f != null) f.paint else Paint())
    }

    fun setClipPath(p: Path) {
        this.p = p
    }

    fun init(c: Canvas) {
        if (p != null) {
            c.clipPath(p!!)
        }
    }

    companion object {

        val FUNCTION_CLEAR: Byte = 0
        val FUNCTION_AND: Byte = 1
        val FUNCTION_AND_REVERSE: Byte = 2
        val FUNCTION_COPY: Byte = 3
        val FUNCTION_AND_INVERTED: Byte = 4
        val FUNCTION_NOOP: Byte = 5
        val FUNCTION_XOR: Byte = 6
        val FUNCTION_OR: Byte = 7
        val FUNCTION_NOR: Byte = 8
        val FUNCTION_EQUIV: Byte = 9
        val FUNCTION_INVERT: Byte = 10
        val FUNCTION_OR_REVERSE: Byte = 11
        val FUNCTION_COPY_INVERTED: Byte = 12
        val FUNCTION_OR_INVERTED: Byte = 13
        val FUNCTION_NAND: Byte = 14
        val FUNCTION_SET: Byte = 15

        val STYLE_SOLID: Byte = 0
        val STYLE_ON_OFF_DASH: Byte = 1
        val STYLE_DOUBLE_DASH: Byte = 2

        val CAP_STYLE_NOT_LAST: Byte = 0
        val CAP_STYLE_BUTT: Byte = 1
        val CAP_STYLE_ROUND: Byte = 2
        val CAP_STYLE_PROJECTING: Byte = 3

        val JOIN_STYLE_MITER: Byte = 0
        val JOIN_STYLE_ROUND: Byte = 1
        val JOIN_STYLE_BEVEL: Byte = 2

        val FILL_STYLE_SOLID: Byte = 0
        val FILL_STYLE_TILED: Byte = 1
        val FILL_STYLE_STIPPLED: Byte = 2
        val FILL_STYLE_OPAQUE_STIPPLED: Byte = 3

        val FILL_RULE_EVEN_ODD: Byte = 0
        val FILL_RULE_WINDING: Byte = 1

        val SUBWINDOW_MODE_CLIP_BY_CHILDREN: Byte = 0
        val SUBWINDOW_MODE_INCLUDE_INFERIORS: Byte = 1

        val ARC_MODE_CHORD: Byte = 0
        val ARC_MODE_PIE_SLICE: Byte = 1
    }
}
